### PR TITLE
[MM-68463] New endpoint to GET user by auth_data

### PR DESCRIPTION
**Pull request by BenCookie95**
## Summary

Adds GET `/api/v4/users/auth_data?value={auth_data}` (system-admin / local-mode only). Given a user's auth data value, the endpoint returns the matching user — including their AuthService, TermsOfService info, and the rest of the standard sanitized user payload.

Backed by a new store method User().GetByAuthData(authData) that resolves a user by AuthData alone. The existing two-argument User().GetByAuth(authData, authService) is left untouched.

### Why a single-parameter lookup
Users.AuthData has a global, column-level UNIQUE constraint ([000046_create_users.up.sql](vscode-file://vscode-app/Applications/Cursor.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/server/channels/db/migrations/postgres/000046_create_users.up.sql) line 8) that is not qualified by AuthService. The codebase has implicitly relied on this for years:

- User.IsValid requires AuthData to come paired with an AuthService and forbids combining AuthData with Password, so any non-null AuthData belongs to exactly one user with exactly one auth service.
- SqlUserStore.UpdateAuthData already maps the DB unique-constraint violation on users_authdata_key to a structured error — the store explicitly treats the column as globally unique.
- Existing enterprise callers (SAML, LDAP, Intune) pass AuthService to GetByAuth not as a uniqueness disambiguator but as a service-context guard ("I'm in the SAML flow, only match SAML users"). The schema makes the disambiguation case impossible.

The legacy GetByAuth(authData, authService) is intentionally preserved for the SSO login flows where the service-scope guard is desirable.

### Caveats
- I had to use a query parameter for the authdata value because of the objectGUID case where we store ObjectGUID values like: `\61\14\e1\d1\c5\35\18\4a\b6\60\d6\78\50\fd\0d\5d`. URL parameters would have issues with stripping leading backslashes and other random encoding issues that don't happen with query parameters.
- AD returns objectGUID as 16 raw bytes inside the `LDAP.SearchResultEntry`. The wire bytes are e.g. `61 14 e1 d1 c5 35 18 4a b6 60 d6 78 50 fd 0d 5d`. MM grabs the raw []byte and runs `getADLdapIdQueryFormat`, which produces `\61\14\e1\d1\c5\35\18\4a\b6\60\d6\78\50\fd\0d\5d` (48 ASCII chars). That string goes into users.authdata as the user identifier. All LDAP providers display the normalized ObjectGUID value in a different way, so if an admin uses a copy of ObjectGUID from `PowerShell` they will see a UUID which obviously won't match to any Mattermost user if you use this new endpoint.

## Ticket
https://mattermost.atlassian.net/browse/MM-68463

<div><a href="https://cursor.com/agents/bc-24799289-e2da-4b9c-a6f1-d4e1a72d9c6d"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-web-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-web-light.png">Open in Web</picture></a>&nbsp;<a href="https://cursor.com/background-agent?bcId=bc-24799289-e2da-4b9c-a6f1-d4e1a72d9c6d"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-cursor-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-cursor-light.png">Open in Cursor</picture></a>&nbsp;</div>

## Release Note

```release-note
New endpoint to fetch users by their auth_data GET `/api/v4/users/auth_data?value={auth_data}`. Only available to sysadmins.
```

## Change Impact: 🟡 Medium

**Regression Risk:** The changes introduce a new sysadmin-only user lookup endpoint and add a GetByAuthData path through store, service, retry/timer layers, and client surfaces; these are additive but touch authentication lookup flows and shared store interfaces, so regression risk is moderate.  
**QA Recommendation:** Run focused QA on permission enforcement (sysadmin-only), query-parameter edge cases (empty/oversized/escaped/special characters including LDAP objectGUID formats), ETag/cache behavior, TermsOfService population, and 404/error mappings; automated tests exist but targeted manual verification is advised.

*Generated by CodeRabbitAI*

[Review Change Stack](https://app.coderabbit.ai/change-stack/mattermost/mattermost/pull/36352)

[Review Change Stack](https://app.coderabbit.ai/change-stack/mattermost/mattermost/pull/36352)

### CODE REVIEW

File: server/channels/api4/user.go [hunk 1]
```diff
@@ -461,6 +462,72 @@ func getUserByEmail(c *Context, w http.ResponseWriter, r *http.Request) {
...
+			c.Err = err2
+			return
+		}
+		if restrictions != nil {
+			c.SetPermissionError(model.PermissionViewMembers)
+			return
+		}
+		c.Err = err
+		return
+	}
```
Comment (cpoile) [inline server/channels/api4/user.go:483]: I think this is unreachable since the user has to be admin, right?

File: server/channels/api4/user.go [hunk 2]
```diff
@@ -461,6 +462,72 @@ func getUserByEmail(c *Context, w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func getUserByAuth(c *Context, w http.ResponseWriter, r *http.Request) {
```
Comment (cpoile) [inline server/channels/api4/user.go]: I wonder if this shouldn't be `getUserByAuthData` / `localGetUserByAuthData` to match the app, store, client layers?

File: server/channels/api4/user.go [hunk 3]
```diff
@@ -461,6 +462,72 @@ func getUserByEmail(c *Context, w http.ResponseWriter, r *http.Request) {
...
+		c.Err = err2
+		return
+	}
+
+	if !canSee {
+		c.SetPermissionError(model.PermissionViewMembers)
+		return
+	}
+
+	if c.IsSystemAdmin() || c.AppContext.Session().UserId == user.Id {
```
Comment (cpoile) [inline server/channels/api4/user.go]: nit: is isSystemAdmin needed?

**Review by cpoile — CHANGES_REQUESTED** [review]
Just some thoughts, let me know what you think. Thanks Ben!

**Review by cpoile — APPROVED** [review]

**Comment by BenCookie95** [comment]
@coderabbitai resolve

**Comment by BenCookie95** [comment]
@DHaussermann to test post-merge. Wanted to get this merged on time for v11.8 feature complete
